### PR TITLE
customize congrats page for cartoon network by leveraging existing company profile infrastructure.

### PR DESCRIPTION
**Pull request by davidsbailey**
cartoon network integration take 2. This is a much hackier but minimizes impact on other codepaths.
### screenshots

see https://github.com/code-dot-org/code-dot-org/pull/5860
### manual verification steps
1. CN message shows in Arabic when starting from CN link then going to congrats page
   - https://code.org/api/hour/begin/gumball?company=CN&lang=ar
   - get redirected to studio.code.org/s/gumball/stage/1/puzzle/1 in arabic
   - click "I have finished the Hour of Code"
   - --> Congrats page shows message in Arabic
2. CN message shows in English when changing language to English
   - https://code.org/api/hour/begin/gumball?company=CN&lang=ar
   - get redirected to studio.code.org/s/gumball/stage/1/puzzle/1 in arabic
   - change language to English
   - click "I have finished the Hour of Code"
   - --> Congrats page shows message in English
3. CN message does not show after completing another tutorial (hard to test locally)
   - https://code.org/api/hour/begin/gumball?company=CN&lang=ar
   - get redirected to studio.code.org/s/gumball/stage/1/puzzle/1 in arabic
   - https://code.org/api/hour/begin/frozen
   - get redirected to studio.code.org/s/frozen/stage/1/puzzle/1 in Arabic
   - click "I have finished the Hour of Code"
   - --> Congrats page does not show CN message
4. CN message does not show after completing another tutorial, even if that tutorial is gumball
   - https://code.org/api/hour/begin/gumball?company=CN&lang=ar
   - get redirected to studio.code.org/s/gumball/stage/1/puzzle/1 in arabic
   - https://code.org/api/hour/begin/gumball
   - get redirected to studio.code.org/s/gumball/stage/1/puzzle/1 in Arabic
   - click "I have finished the Hour of Code"
   - --> Congrats page does not show CN message

### CODE REVIEW

File: pegasus/routes/hoc_routes.rb [hunk 1]
```diff
@@ -25,10 +25,19 @@
...
     company = nil unless DB[:forms].where(kind: 'CompanyProfile', name: company).first
   end
 
+  # The lang parameter is used only by the cartoon network integration.
+  lang = request.GET['lang']
+  if company == 'CN' && lang == 'ar'
+    response.set_cookie('language_', {value: lang, domain: ".code.org", path: '/', expires: Time.now + (365*24*3600)})
+  end
+
   launch_tutorial(tutorial, company: company)
```
Comment (laurelfan) [inline pegasus/routes/hoc_routes.rb:40]: what is launch_tutorial and how does it behave with companies with no company profiule?

File: pegasus/routes/hoc_routes.rb [hunk 2]
```diff
@@ -25,10 +25,19 @@
 
   # set company to nil if not a valid company
   company = request.GET['company']
-  unless company.nil?
+  # Our partnership with cartoon network piggy-backs on our company profile.
+  # Pass their company param through to the congrats page even if no entry exists
+  # in the forms table.
+  unless company.nil? || company == 'CN'
```
Comment (philbogle) [inline pegasus/routes/hoc_routes.rb]: "Pass through the company param to the congrats page only if an entry exists in the forms or for the special case of cartoon network, where we need the company to add a customized link"

File: pegasus/sites.v3/code.org/public/congrats.haml [hunk 3]
```diff
@@ -26,7 +26,10 @@ max_age: 60
 -twitter[:hashtags] = 'HourOfCode' unless I18n.t(:just_did_hoc).include? '#HourOfCode'
 
 -company_param = params[:co]
--form = DB[:forms].where(kind:'CompanyProfile', name:company_param).first if company_param
+-# Our partnership with cartoon network piggy-backs on our company profile infrastructure.
```
Comment (philbogle) [inline pegasus/sites.v3/code.org/public/congrats.haml]: "Do not show company profile sections for the cartoon network, even if someone accidentally creates an entry for them in the forms table.  (Our partnership with cartoon network piggy-backs on our company profile infrastructure.)"

File: pegasus/routes/hoc_routes.rb (hunk 1 above)
Comment (davidsbailey) [inline pegasus/routes/hoc_routes.rb:40]: launch tutorial sets the hour of code cookie and then redirects you to studio.code.org/s/<tutorial>/reset (for our tutorials), or to an external site.

```
def launch_tutorial(tutorial,params={})
  unless settings.read_only || hoc_activity_writes_disabled
    row = create_session_row(
      referer: request.referer_site_with_port,
      tutorial: tutorial[:code],
      company: params[:company],
      started_at: DateTime.now,
      started_ip: request.ip,
    )
    set_hour_of_code_cookie_for_row(row)
  end

  dont_cache
  redirect tutorial[:url], 302
end
```

File: pegasus/routes/hoc_routes.rb [hunk 4]
```diff
@@ -25,10 +25,19 @@
 
   # set company to nil if not a valid company
   company = request.GET['company']
-  unless company.nil?
+  # Our partnership with cartoon network piggy-backs on our company profile.
```
Comment (philbogle) [inline pegasus/routes/hoc_routes.rb]: Define a constant for CN:

CARTOON_NETWORK = 'CN'

File: pegasus/sites.v3/code.org/public/congrats.haml [hunk 5]
```diff
@@ -49,10 +52,18 @@ max_age: 60
   -else
     %h1=I18n.t :congratulations_on_completing_hoc
 
-  -if request.locale == 'ar-SA'
-    %h2
-      أنهيت ساعة البرمجة ! العودة إلى
-      %a{:href=>"http://click-and-share.com/hour_of_code/data-capture/"} كرتون نتورك
+
+  -if company_param == 'CN' && cert_script == 'gumball'
+    -if request.locale == 'ar-SA'
```
Comment (philbogle) [inline pegasus/sites.v3/code.org/public/congrats.haml:58]: Do we really want ar-SA specifically or just 'ar-*'

**Comment by philbogle** [comment]
LGTM after comments are addressed and you add the unit test.

File: pegasus/sites.v3/code.org/public/congrats.haml (hunk 5 above)
Comment (davidsbailey) [inline pegasus/sites.v3/code.org/public/congrats.haml:58]: @tanyaparker confirmed we only want ar-SA here.

**Pull request merged**